File: apps/desktop/src/contracts/runtimeStatus.ts
export const RUNTIME_STATUS_PROTOCOL = "hive-runtime-status-ipc-v1" as const;
export const RUNTIME_STATUS_SCHEMA = "hive-runtime-status-v1" as const;
export const MAX_RUNTIME_STATUS_RESPONSE_BYTES = 40_000;

export type RuntimeState = "READY" | "UNKNOWN" | "DISCONNECTED" | "DEGRADED";

export interface RuntimeStatusEnvelope {
  protocol: typeof RUNTIME_STATUS_PROTOCOL;
  requestId: string;
  ok: true;
  snapshot: {
    schema: typeof RUNTIME_STATUS_SCHEMA;
    runtime: { state: RuntimeState; provenance: string; detail: string };
    providers: Array<{ providerId: string; modelIds: string[]; state: RuntimeState }>;
    task: null | { taskId: string; state: string; nodeTotal: number; nodeSucceeded: number; executions: number; failures: number };
    permission: { state: RuntimeState; policyEpoch: number | null; activeSessions: number | null; pendingApprovals: number | null };
  };
}

const STATES = new Set<RuntimeState>(["READY", "UNKNOWN", "DISCONNECTED", "DEGRADED"]);

function record(value: unknown, keys: string[]): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) throw new Error("expected object");
  const actual = Object.keys(value as Record<string, unknown>).sort();
  const expected = [...keys].sort();
  if (actual.length !== expected.length || actual.some((key, index) => key !== expected[index])) throw new Error("unexpected object shape");
  return value as Record<string, unknown>;
}

function text(value: unknown, limit: number): string {
  if (typeof value !== "string" || value.trim().length < 1 || value.trim().length > limit) throw new Error("invalid bounded text");
  return value.trim();
}

function state(value: unknown): RuntimeState {
  if (typeof value !== "string" || !STATES.has(value as RuntimeState)) throw new Error("invalid state");
  return value as RuntimeState;
}

function counter(value: unknown): number {
  if (!Number.isInteger(value) || (value as number) < 0 || (value as number) > 1_000_000) throw new Error("invalid counter");
  return value as number;
}

function nullableCounter(value: unknown): number | null {
  return value === null ? null : counter(value);
}

export function parseRuntimeStatusEnvelope(input: unknown): RuntimeStatusEnvelope {
  const bytes = new TextEncoder().encode(JSON.stringify(input)).byteLength;
  if (bytes > MAX_RUNTIME_STATUS_RESPONSE_BYTES) throw new Error("runtime status response exceeds byte ceiling");
  const root = record(input, ["protocol", "requestId", "ok", "snapshot"]);
  if (root.protocol !== RUNTIME_STATUS_PROTOCOL || root.ok !== true) throw new Error("unsupported runtime status envelope");
  const requestId = text(root.requestId, 64);
  if (!/^[A-Za-z0-9_.:-]+$/.test(requestId)) throw new Error("invalid request id");
  const snapshot = record(root.snapshot, ["schema", "runtime", "providers", "task", "permission"]);
  if (snapshot.schema !== RUNTIME_STATUS_SCHEMA) throw new Error("unsupported runtime status schema");
  const runtime = record(snapshot.runtime, ["state", "provenance", "detail"]);
  const runtimeValue = { state: state(runtime.state), provenance: text(runtime.provenance, 160), detail: text(runtime.detail, 160) };
  if (!Array.isArray(snapshot.providers) || snapshot.providers.length > 16) throw new Error("invalid providers");
  const seenProviders = new Set<string>();
  const providers = snapshot.providers.map((raw) => {
    const provider = record(raw, ["providerId", "modelIds", "state"]);
    const providerId = text(provider.providerId, 80).toLowerCase();
    if (seenProviders.has(providerId)) throw new Error("duplicate provider");
    seenProviders.add(providerId);
    if (!Array.isArray(provider.modelIds) || provider.modelIds.length > 64) throw new Error("invalid models");
    const modelIds = provider.modelIds.map((model) => text(model, 128));
    if (new Set(modelIds).size !== modelIds.length) throw new Error("duplicate model");
    const providerState = state(provider.state);
    if (providerState === "READY" && modelIds.length === 0) throw new Error("fake provider readiness");
    return { providerId, modelIds, state: providerState };
  });
  let task: RuntimeStatusEnvelope["snapshot"]["task"] = null;
  if (snapshot.task !== null) {
    const rawTask = record(snapshot.task, ["taskId", "state", "nodeTotal", "nodeSucceeded", "executions", "failures"]);
    const nodeTotal = counter(rawTask.nodeTotal);
    const nodeSucceeded = counter(rawTask.nodeSucceeded);
    const executions = counter(rawTask.executions);
    const failures = counter(rawTask.failures);
    if (nodeTotal > 256 || nodeSucceeded > nodeTotal || failures > executions) throw new Error("inconsistent task counters");
    task = { taskId: text(rawTask.taskId, 128), state: text(rawTask.state, 32), nodeTotal, nodeSucceeded, executions, failures };
  }
  const rawPermission = record(snapshot.permission, ["state", "policyEpoch", "activeSessions", "pendingApprovals"]);
  const permissionState = state(rawPermission.state);
  const policyEpoch = nullableCounter(rawPermission.policyEpoch);
  const activeSessions = nullableCounter(rawPermission.activeSessions);
  const pendingApprovals = nullableCounter(rawPermission.pendingApprovals);
  if ((permissionState === "UNKNOWN" || permissionState === "DISCONNECTED") && [policyEpoch, activeSessions, pendingApprovals].some((value) => value !== null)) {
    throw new Error("unknown permission state carries counters");
  }
  return {
    protocol: RUNTIME_STATUS_PROTOCOL,
    requestId,
    ok: true,
    snapshot: {
      schema: RUNTIME_STATUS_SCHEMA,
      runtime: runtimeValue,
      providers,
      task,
      permission: { state: permissionState, policyEpoch, activeSessions, pendingApprovals },
    },
  };
}

File: hive_runtime/runtime_status.py
"""Bounded non-authoritative runtime observability for Hive Coder.

This module intentionally exports presentation state only. It cannot mint permits,
approve requests, execute providers, run prompts, mutate tasks, or expose credentials.
"""
from __future__ import annotations

import json
from dataclasses import dataclass
from enum import Enum
from typing import Iterable, Mapping

from .agent_tasks import TaskSnapshot
from .providers import ProviderCatalog

STATUS_SCHEMA = "hive-runtime-status-v1"
MAX_STATUS_BYTES = 32_768
MAX_TEXT = 160
MAX_PROVIDERS = 16
MAX_MODELS_PER_PROVIDER = 64
MAX_TASK_NODES = 256


class StatusState(str, Enum):
    READY = "READY"
    UNKNOWN = "UNKNOWN"
    DISCONNECTED = "DISCONNECTED"
    DEGRADED = "DEGRADED"


def _bounded_text(value: object, *, field: str, limit: int = MAX_TEXT) -> str:
    if not isinstance(value, str):
        raise ValueError(f"{field} must be a string")
    text = value.strip()
    if not text or len(text) > limit:
        raise ValueError(f"{field} must contain 1..{limit} characters")
    return text


@dataclass(frozen=True)
class StatusSignal:
    state: StatusState
    provenance: str
    detail: str

    def validated(self) -> "StatusSignal":
        _bounded_text(self.provenance, field="provenance")
        _bounded_text(self.detail, field="detail")
        return self


@dataclass(frozen=True)
class ProviderStatus:
    provider_id: str
    model_ids: tuple[str, ...]
    state: StatusState = StatusState.READY

    def validated(self) -> "ProviderStatus":
        _bounded_text(self.provider_id, field="provider_id", limit=80)
        if len(self.model_ids) > MAX_MODELS_PER_PROVIDER:
            raise ValueError("too many provider models")
        seen: set[str] = set()
        for model_id in self.model_ids:
            normalized = _bounded_text(model_id, field="model_id", limit=128)
            if normalized in seen:
                raise ValueError("duplicate provider model id")
            seen.add(normalized)
        if self.state is StatusState.READY and not self.model_ids:
            raise ValueError("READY provider requires at least one observed model")
        return self


@dataclass(frozen=True)
class TaskStatusSummary:
    task_id: str
    state: str
    node_total: int
    node_succeeded: int
    executions: int
    failures: int

    def validated(self) -> "TaskStatusSummary":
        _bounded_text(self.task_id, field="task_id", limit=128)
        _bounded_text(self.state, field="task_state", limit=32)
        counters = (self.node_total, self.node_succeeded, self.executions, self.failures)
        if any(not isinstance(value, int) or value < 0 for value in counters):
            raise ValueError("task counters must be non-negative integers")
        if self.node_total > MAX_TASK_NODES or self.node_succeeded > self.node_total:
            raise ValueError("task node counters out of bounds")
        if self.failures > self.executions:
            raise ValueError("task failure count exceeds executions")
        return self


@dataclass(frozen=True)
class PermissionStatusSummary:
    state: StatusState
    policy_epoch: int | None
    active_sessions: int | None
    pending_approvals: int | None

    def validated(self) -> "PermissionStatusSummary":
        for name, value in (
            ("policy_epoch", self.policy_epoch),
            ("active_sessions", self.active_sessions),
            ("pending_approvals", self.pending_approvals),
        ):
            if value is not None and (not isinstance(value, int) or value < 0 or value > 1_000_000):
                raise ValueError(f"{name} is out of bounds")
        if self.state in {StatusState.UNKNOWN, StatusState.DISCONNECTED} and any(
            value is not None for value in (self.policy_epoch, self.active_sessions, self.pending_approvals)
        ):
            raise ValueError("unknown/disconnected permission state cannot carry authoritative-looking counters")
        return self


@dataclass(frozen=True)
class RuntimeStatusSnapshot:
    runtime: StatusSignal
    providers: tuple[ProviderStatus, ...]
    task: TaskStatusSummary | None
    permission: PermissionStatusSummary
    schema: str = STATUS_SCHEMA

    def validated(self) -> "RuntimeStatusSnapshot":
        if self.schema != STATUS_SCHEMA:
            raise ValueError("unsupported runtime status schema")
        self.runtime.validated()
        if len(self.providers) > MAX_PROVIDERS:
            raise ValueError("too many providers")
        seen: set[str] = set()
        for provider in self.providers:
            provider.validated()
            key = provider.provider_id.strip().lower()
            if key in seen:
                raise ValueError("duplicate provider id")
            seen.add(key)
        if self.task is not None:
            self.task.validated()
        self.permission.validated()
        encoded = json.dumps(self.to_dict_unchecked(), sort_keys=True, separators=(",", ":")).encode("utf-8")
        if len(encoded) > MAX_STATUS_BYTES:
            raise ValueError("runtime status exceeds byte ceiling")
        return self

    def to_dict_unchecked(self) -> dict[str, object]:
        return {
            "schema": self.schema,
            "runtime": {
                "state": self.runtime.state.value,
                "provenance": self.runtime.provenance,
                "detail": self.runtime.detail,
            },
            "providers": [
                {
                    "providerId": provider.provider_id,
                    "modelIds": list(provider.model_ids),
                    "state": provider.state.value,
                }
                for provider in self.providers
            ],
            "task": None
            if self.task is None
            else {
                "taskId": self.task.task_id,
                "state": self.task.state,
                "nodeTotal": self.task.node_total,
                "nodeSucceeded": self.task.node_succeeded,
                "executions": self.task.executions,
                "failures": self.task.failures,
            },
            "permission": {
                "state": self.permission.state.value,
                "policyEpoch": self.permission.policy_epoch,
                "activeSessions": self.permission.active_sessions,
                "pendingApprovals": self.permission.pending_approvals,
            },
        }

    def to_json(self) -> str:
        self.validated()
        return json.dumps(self.to_dict_unchecked(), sort_keys=True, separators=(",", ":"))


SECRET_KEYS = ("token", "secret", "password", "credential", "api_key", "apikey", "permit", "approval_token")


def reject_secret_like_mapping(values: Mapping[str, object]) -> None:
    """Reject accidental secret-bearing observation input before reduction."""
    for key in values:
        lowered = str(key).strip().lower()
        if any(marker in lowered for marker in SECRET_KEYS):
            raise ValueError("secret-like observation keys are not allowed")


def summarize_provider_catalog(catalog: ProviderCatalog, provider_ids: Iterable[str]) -> tuple[ProviderStatus, ...]:
    summaries: list[ProviderStatus] = []
    for raw_provider in provider_ids:
        provider_id = _bounded_text(raw_provider, field="provider_id", limit=80).lower()
        models = tuple(model.model_id for model in catalog.models(provider_id))
        state = StatusState.READY if models else StatusState.UNKNOWN
        summaries.append(ProviderStatus(provider_id, models, state).validated())
        if len(summaries) > MAX_PROVIDERS:
            raise ValueError("too many providers")
    return tuple(summaries)


def summarize_task(snapshot: TaskSnapshot) -> TaskStatusSummary:
    total = len(snapshot.node_status)
    succeeded = sum(1 for _, state in snapshot.node_status if state.value == "succeeded")
    return TaskStatusSummary(
        task_id=snapshot.task_id,
        state=snapshot.status.value,
        node_total=total,
        node_succeeded=succeeded,
        executions=snapshot.executions,
        failures=snapshot.failures,
    ).validated()


def disconnected_snapshot() -> RuntimeStatusSnapshot:
    return RuntimeStatusSnapshot(
        runtime=StatusSignal(StatusState.DISCONNECTED, "hive-runtime-status", "No trusted live runtime observation is connected."),
        providers=(),
        task=None,
        permission=PermissionStatusSummary(StatusState.DISCONNECTED, None, None, None),
    ).validated()
